internal/pkg/client/cache/dir.go

```diff
@@ -22,69 +22,251 @@ const (
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the Library cache: %s", err)
+	}
+	newCache.OciTemp, err = getOciTempCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the OCI temp cache")
+	}
+	newCache.OciBlob, err = getOciBlobCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the OCI blob cache")
+	}
+	newCache.Net, err = getNetCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the Net cache")
+	}
+	newCache.Shub, err = getShubCachePath(newCache)
+	if err != nil {
+		return nil, fmt.Errorf("failed getting the path to the Shub cache")
+	}
 
-// Root is the root location where all of singularity caching happens. Library, Shub,
-// and oci image formats supported by containers/image repository will be cached inside
-//
-// Defaults to ${HOME}/.singularity/cache
-func Root() string {
-	updateCacheRoot()
+	// Sanity check to ensure that everything is fine before returning the
+	// handle. We do not know a way at the moment to reach the error case.
+	if !newCache.IsValid() {
+		return nil, fmt.Errorf("unable to correctly initialize new cache")
+	}
 
-	return root
+	return newCache, nil
 }
 
-// Clean : wipes all files in the cache directory, will return a error if one occurs
-func Clean() error {
-	sylog.Debugf("Removing: %v", Root())
+// Destroy a specific Singularity cache
+func (c *SingularityCache) Destroy() error {
+	sylog.Debugf("Removing: %v", c.Root)
+	if !c.IsValid() {
+		return fmt.Errorf("invalid cache")
+	}
 
-	if err := os.RemoveAll(Root()); err != nil {
-		return fmt.Errorf("unable to clean all cache: %s", err)
+	err := os.RemoveAll(c.Root)
+	if err != nil {
+		return fmt.Errorf("failed to delete the cache: %s", err)
 	}
 
 	return nil
 }
 
-func updateCacheRoot() {
+// IsValid checks whether a given Singularity cache is valid or not
+func (c *SingularityCache) IsValid() bool {
+	if c.State != StateInitialized {
+		return false
+	}
+
+	if c.BaseDir == "" {
+		return false
+	}
+
+	// Since Clean/Destroy delete everything in the cache directory,
+	// we make sure that when the user set the environment variable
+	// to specify where the cache should be, it cannot be in critical
+	// directory such as $HOME
+	usr, err := user.Current()
+	if err != nil {
+		return false
+	}
+
+	// The root cannot be empty of the Home directory. If the root was
+	// to be the home directory, destroying the cache would delete
+	// user's data
+	if c.Root == "" || c.Root == usr.HomeDir {
+		return false
+	}
+
+	// Basic check of the sub-cache validity
+	if c.Library == "" || c.Net == "" || c.OciTemp == "" || c.OciBlob == "" || c.Shub == "" {
+		return false
+	}
+
+	return true
+}
+
+// Clean wipes all files in the cache directory, will return a error if one occurs
+// Since renamed Destroy() but kept for backward compatibility
+func (c *SingularityCache) Clean() error {
```

Comment:
there's a single use of Clean outside this package, in `internal/app/singularity/cache_clean_linux.go`

For that specific use, the name `Clean` seems appropriate.

Maybe unexport `Destroy`?

Reply:
To me, `Clean()` does not match the actual semantic of the function, which erase everything. I kept both to not be too intrusive (introduce new API -> deprecate old one -> delete old one). If one function would have to not export one, I would not export `Clean()` instead of `Destroy()`.